### ISSUE TITLE
Enhance NIP-01 with IPFS CIDs for Decentralized Media Storage

### ISSUE THREAD

**Issue by besoeasy** [nostr-protocol/nips#1285]
Here is the current NIP1 Object

`{
  "id": "3237fd9a04af4cfcca803cc1825173c44f3ab360439820d30826d2016dfc0783",
  "pubkey": "f680ddd8c13f1ceab92edd3495f0fce747bb80c270e081f80c7401befb62da7c",
  "created_at": 1717795867,
  "kind": 1,
  "tags": [
    [
      "r",
      "https://i.ibb.co/JjYZQVN/graph.png"
    ]
  ],
  "content": "Bitcoin Looking Beautiful 😍 \n\nhttps://i.ibb.co/JjYZQVN/graph.png",
  "sig": "12b47bdf2db31b2630d219cc450b78a2a17a89f209b5fb16c7032c5264735052a3d8cabe1f60ae97957e2c581e07e6808eaa1b391d941a0915d9759825c52604"
}`

All we need to do is to add a "cid" field that resolves to IPFS Hash

**Comment by besoeasy** [comment]
Here is the current NIP1 Object

`{
  "id": "3237fd9a04af4cfcca803cc1825173c44f3ab360439820d30826d2016dfc0783",
  "pubkey": "f680ddd8c13f1ceab92edd3495f0fce747bb80c270e081f80c7401befb62da7c",
  "created_at": 1717795867,
  "kind": 1,
  "tags": [
    [
      "r",
      "https://i.ibb.co/JjYZQVN/graph.png"
    ]
  ],
  "content": "Bitcoin Looking Beautiful 😍 \n\nhttps://i.ibb.co/JjYZQVN/graph.png",
  "sig": "12b47bdf2db31b2630d219cc450b78a2a17a89f209b5fb16c7032c5264735052a3d8cabe1f60ae97957e2c581e07e6808eaa1b391d941a0915d9759825c52604"
}`

**Comment by vitorpamplona** [comment]
I don't understand what you are trying to propose... 

No client has an IPFS node client to pull images from the IPFS network. Are you proposing this becomes a requirement for all clients?

**Comment by besoeasy** [comment]
## Description
Currently, NIP-01 supports decentralized storage for text-based content. However, it would be beneficial to extend this functionality to include multimedia files like images and videos. This can be achieved by utilizing InterPlanetary File System (IPFS) Content Identifiers (CIDs) to store and reference these files.

## Example
1. **NIP-01 Update:** Modify NIP-01 to include a new field for storing IPFS CIDs for multimedia files. This would allow users to upload and reference images and videos directly from their Nostr profiles.
2. **IPFS Integration:** Integrate IPFS into the Nostr protocol to enable decentralized storage and retrieval of multimedia files. This would ensure that files are stored and accessed in a decentralized manner, reducing reliance on centralized servers.

## Benefits
- **Decentralized Storage:** IPFS CIDs would allow for decentralized storage and retrieval of multimedia files, ensuring that no single entity controls the storage or access to these files.
- **Improved Security:** By using IPFS CIDs, the risk of data loss or tampering would be significantly reduced, as the files are stored in a decentralized network.
- **Enhanced User Experience:** Users would have more flexibility in how they store and share multimedia content, making it easier to collaborate and interact with others.

## Steps to Implement
1. **Update NIP-01:** Modify NIP-01 to include the new field for storing IPFS CIDs.
2. **Integrate IPFS:** Integrate IPFS into the Nostr protocol to enable decentralized storage and retrieval of multimedia files.

**Comment by vitorpamplona** [comment]
Why not just place it as a tag?

**Comment by besoeasy** [comment]
> Why not just place it as a tag?
 
There are several reasons why using IPFS CIDs as a native reference for multimedia content is more beneficial than simply placing it as a tag:

### 1. Native Support in Browsers and OS
IPFS is natively supported in some mainstream browsers like Brave and many Linux-based operating systems. This means that users can directly access IPFS content without needing additional software or plugins. By using IPFS CIDs, we can leverage this native support to provide a seamless user experience.

### 2. Decentralized Caching
One of the significant advantages of IPFS is that the entire network caches the content. This means that even if a specific gateway goes down, users can switch to another gateway and still access the content. In contrast, if a centralized server like nostr.build goes down, a significant portion of content images would become unavailable.

### 3. Flexibility and Portability
IPFS CIDs can be accessed through various gateways if native support is missing. This ensures that users can always access the content, even if their browser or OS does not natively support IPFS. The use of IPFS CIDs provides a flexible and portable solution for decentralized media storage.

### 4. Standardization
IPFS CIDs follow a standardized format, making it easier for different platforms and applications to integrate and support them. This standardization also enables better interoperability across different systems.

### 5. Security and Integrity
IPFS CIDs are based on the cryptographic hash of the content, ensuring that any changes to the content result in a different CID. This provides a robust mechanism for verifying the integrity and authenticity of the content.

In summary, using IPFS CIDs as a native reference for multimedia content offers several benefits, including native support in browsers and OS, decentralized caching, flexibility, standardization, and security. These advantages make IPFS CIDs a more robust and reliable solution compared to simply placing them as tags.

I been a experienced IPFS developer, i can show a small benchmark if the community wants and really proove that IPFS as media CDN would solve our dependencies on Centerlised Image Hosters

**Comment by besoeasy** [comment]
I forgot to add that IPFS, supports, all content types

From Images to direct .torrent files

Everything described in all of our NIPs is supported by IPFS

**Comment by vitorpamplona** [comment]
Why are you posting things from ChatGPT? 

We know what IPFS is. That's not the question. The question is how are you going to integrate it. Changing NIP-01 is not needed to support IPFS.

**Comment by alexgleason** [comment]
The way to integrate IPFS with Nostr is easy. Just include a URL like `https://<gateway>/ipfs/<cid>`

This is backwards and forwards compatible. I've been doing it on Ditto: https://gitlab.com/soapbox-pub/ditto/-/blob/main/src/uploaders/IPFSUploader.ts?ref_type=heads

**Comment by besoeasy** [comment]
i am not using chatgpt, i am using a translator AI, i do not speak english

> We know what IPFS is. That's not the question. The question is how are you going to integrate it. Changing NIP-01 is not needed to support IPFS.

i understand your question, i will post comparison, as soon i get some time, showing problems with NIP 1 Object

i will mention you here when i am done !

**Comment by besoeasy** [comment]
> The way to integrate IPFS with Nostr is easy. Just include a URL like `https://<gateway>/ipfs/<cid>`
> 
> This is backwards and forwards compatible. I've been doing it on Ditto: https://gitlab.com/soapbox-pub/ditto/-/blob/main/src/uploaders/IPFSUploader.ts?ref_type=heads

we need to implement it as uri ipfs:// or "cid" field 

to get rid of dependencies on gateways, if ipfs content is hardcoded with gateways, it makes it no difference then nostr.build or imgbb

**Comment by alexgleason** [comment]
The IPFS path gateway is standardized: https://docs.ipfs.tech/concepts/ipfs-gateway/#gateway-types

This is superior to ipfs:// URIs because it works on the legacy web, AND ipfs enabled tools can resolve it.

It is the same as an ipfs URI except it contains extra information. There is no reason not to use this.

**Comment by besoeasy** [comment]
> The IPFS path gateway is standardized: https://docs.ipfs.tech/concepts/ipfs-gateway/#gateway-types
> 
> This is superior to ipfs:// URIs because it works on the legacy web, AND ipfs enabled tools can resolve it.
> 
> It is the same as an ipfs URI except it contains extra information. There is no reason not to use this.

**example :**

ipfs://QmRhZLf2ccHag8QLNEf8Jws4KXbtA4bx5K6xLf4WLa4kdn

and 

https://bafybeibr5yx2gie53vmsgptybfpxfc6ykf5polikw3azhctygs2vfszcre.ipfs.dweb.link/

Serve same content, with different methods 

My long term vision was to have ipfs nodes relay pools, just like nostr relays or maybe a combination of both

where node runners can specify how much of nostr and ipfs data they want to cache

its not necessity at the moment, its a improvement that will help NOSTR in long term

+ having IPFS integration standardised will help users self host content locally on computer without setting up a always on https:.. server 

imo this makes NOSTR publicly more appealing 


NOSTR is still new we can implement this, but in 2-3 years from now, this integration will be hard, and it would be easier for us to code and have our own toned down implemetation based on IPFS

**Comment by alexgleason** [comment]
It sounds like what you want is Blossom. https://github.com/hzrd149/blossom

**Comment by besoeasy** [comment]
> It sounds like what you want is Blossom. https://github.com/hzrd149/blossom

doesn't sound decentralised, and requires always on server i assume

**Comment by vitorpamplona** [comment]
The http url renders great on Amethyst. :) 

![Screenshot 2024-06-07 at 6 28 18 PM](https://github.com/nostr-protocol/nips/assets/532031/045a3ce6-8906-4bad-b8f6-486677e7ed4a)

**Comment by besoeasy** [comment]
> The http url renders great on Amethyst. :)
> 
> ![Screenshot 2024-06-07 at 6 28 18 PM](https://private-user-images.githubusercontent.com/532031/337809474-045a3ce6-8906-4bad-b8f6-486677e7ed4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc3OTk2MjUsIm5iZiI6MTcxNzc5OTMyNSwicGF0aCI6Ii81MzIwMzEvMzM3ODA5NDc0LTA0NWEzY2U2LTg5MDYtNGJhZC1iOGY2LTQ4NjY3N2U3ZWQ0YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYwN1QyMjI4NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZTg2ZmQyMzE0MjkzMDRkNzZmZGIyMmM1Y2QyZWIwNGQ3ZGVmM2NlOTJmODhiYzE2Yjc2YjVlNTJhMDdlYjlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.i7C38S6WiRyG-WGV8W1D3hpRPIoquKMm2ajspH1W_6U)

Exactly, it renders good as long as gateway is up


see this 
![image](https://github.com/nostr-protocol/nips/assets/8281782/f645f975-739d-4df5-9f9d-fbffe0329de7)

gateway goes up and down, and a good gateway for u, will not be best option for me

**Comment by vitorpamplona** [comment]
Is there an Android app that serves IPFS images as a full node? If so, then we could just create a setting to load the image from the local app node instead of coding everything in Nostr.

**Comment by besoeasy** [comment]
> Is there an Android app that serves IPFS images as a full node? If so, then we could just create a setting to load the image from the local app node instead of coding everything in Nostr.

**you know what thats smart, i really like this approach**

I am afraid this can only done for non android and ios devices for now, but it is possible, IPFS does have pure JS implementation

**Comment by besoeasy** [comment]
I will start a pilot with **ipfs://** uri **_without changing NIP 01_** as suggest by @vitorpamplona 

i will talk to few clients devs to see if they can support this and see how it works, but i am afraid in order for this to work, we need to adopt this together 

but something tells me, its not going to work well in full decentralised way unless we natively use IPFS and let them handle our troubles with files management & p2p delivery

**Comment by vitorpamplona** [comment]
Just develop a full IPFS client for Android first. Until that exists and becomes extremely reliable, it's useless to debate how to integrate with Nostr.

**Comment by melvincarvalho** [comment]
No changes needed in NIP-01 for this.  If you want to add an IPFS URI such as ifps:// you can shoe-horn it into a tag.  If you want to reference a CID but not be tied to IPFS transport, and use some other discovery method, I'd suggest using `urn:cid:<CID>` which decouples the hash, the transport and the discovery.  Of course, you'd then need to figure out how to do discovery, perhaps using relays and lookups, but that's another NIP.

**Comment by besoeasy** [comment]
> The point of this long reply isn't to say that IPFS wont work or isn't a good system for censorship resistant media. but simply that we don't need a system as complicated or heavy as IPFS to get censorship resistant media, we just need universal file IDs and servers that serve them by those IDs

this is what i am trying to commend, either IPFS or our own in house system

**Comment by besoeasy** [comment]
Maybe easier would be to introduce a new tag in https://github.com/nostr-protocol/nips?tab=readme-ov-file#standardized-tags

["ipfs", "img", "IPFS CID"]
["ipfs", "vid", "IPFS CID"]
["ipfs", "other", "IPFS CID"]

This way it doesn't modify NIP 1, and the same time have IPFS standard for anyone looking to implement it

i see it as win-win for both sides of discussions

**Comment by alexgleason** [comment]
You're overthinking this. Just upload to a public gateway like I said and use a regular URL.

**Comment by fiatjaf** [comment]
If IPFS worked as advised we wouldn't need Nostr at all, so we should either close this or close the entirety of Nostr.
